algorithms/src/MarchingCubes.cpp

```diff
@@ -0,0 +1,145 @@
+/**
+ * @file MarchingCubes.cpp
+ * @author Ihor Lytvyn (litvinigor98@gmail.com)
+ * @date Created May 26, 2019
+ **/
+
+#include "MarchingCubes.h"
+#include "MarchingCubesConfig.h"
+
+#include <vector>
+
+namespace SPHAlgorithms
+{
+
+std::vector<Point3F> MarchingCubes::getFunctionMesh(std::function<float(float, float, float)> f)
+{
+    std::vector<Point3F> resultMesh;
+
+    for (float iX = X_MIN; iX < X_MAX; iX += GRID_CUBE_SIZE)
+    {
+        for (float iY = Y_MIN; iY < Y_MAX; iY += GRID_CUBE_SIZE)
+        {
+            for (float iZ = Z_MIN; iZ < Z_MAX; iZ += GRID_CUBE_SIZE)
+            {
+                // iX, iY, iZ are coordinates of the current vertex in grid_cube
+                auto vertices = MarchCube(f, iX, iY, iZ);
+                std::copy(vertices.begin(), vertices.end(), std::back_inserter(resultMesh));
+            }
+        }
+    }
+    return resultMesh;
+}
+
+static float adapt(float fValue1, float fValue2)
+{
+    const float fDelta = fValue2 - fValue1;
+
+    if (std::abs(fDelta) < 0.0001f)
+    {
+        return 0.5f;
+    }
+
+    return -fValue1 / fDelta;
+}
+
+std::vector<Point3F> MarchingCubes::MarchCube(std::function<float(float, float, float)> f, float fX, float fY, float fZ)
+{
+    // vector with all triangles found
+    std::vector<Point3F> resultTrianglesMesh;
+
+    float CubeValue[8];
+
+    // Evaluate value of the cube vertex at each point
+    for (int iVertex = 0; iVertex < 8; iVertex++)
+    {
+        CubeValue[iVertex] =
+            f(fX + VertexOffset[iVertex][0], fY + VertexOffset[iVertex][1], fZ + VertexOffset[iVertex][2]);
+    }
+
+    // Flag that indicates that vertice is inside or outsice of the surface
+    int iFlagIndex = 0;
+
+    // Find which vertices are inside of the surface and which are outside
+    iFlagIndex = determineFlag(iFlagIndex, CubeValue);
+
+    // Find which edges are intersected by the surface
+    int iEdgeFlags = CubeEdgeFlags[iFlagIndex];
+
+    // If the cube is entirely inside or outside of the surface, then there will be no intersections
+    if (iEdgeFlags == 0)
+    {
+        return resultTrianglesMesh;
+    }
+
+    // Points of intersection of the surface with each edge
+    std::vector<Point3F> EdgeVertex(12);
+
+    // Find points of intersection on each edge
+    findPointIntersection(EdgeVertex, iEdgeFlags, CubeValue, fX, fY, fZ);
+
+    // Fill the triangles that were found.  There can be up to five per cube
+    fillFoundTriangles(resultTrianglesMesh, EdgeVertex, iFlagIndex);
```

Comment:
Should return `resultTrianglesMesh`

Reply:
i dont think that it should return something